File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        int start = 0;
        int max = 0;
        HashSet<Character>set = new HashSet<>();
        
        for(int end = 0; end < s.length(); end++){
            
            while(end < s.length()){
            if(!set.contains(s.charAt(end))){
                set.add(s.charAt(end));
                end++;
                max = Math.max(set.size(), max);
            }
            else{
            set.remove(s.charAt(start));
            start++;
            }
        }
    }
        return max;
    }
}

// Input: s = "pwwkew"
// Output: 3

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        left = 0
        seen = set()
        max_len = 0

        for right in range(len(s)):
            while s[right] in seen:
                seen.remove(s[left])
                left = left + 1
            seen.add(s[right])
            max_len = max(max_len, right - left+1)

        return max_len

        